src/utils/Votes.sol

```diff
@@ -86,6 +86,19 @@ abstract contract Votes is IVotes, IVotesToken {
         handleDelegation(delegator, delegatee);
     }
 
+    /**
+     * @notice clears the delegation of the sender
+     */
+    function clearDelegation() external {
+        address currentDelegate = Checkpoints.delegates(msg.sender);
+        if(currentDelegate == msg.sender) {
+            Checkpoints.moveDelegation(currentDelegate, address(0), getVotes(msg.sender));
+        } else {
+            Checkpoints.moveDelegation(currentDelegate, address(0), IVotesToken(this).balanceOf(msg.sender));
```

Comment:
For my understanding, what happens in this case:

- wallet A delegates their voting power to wallet B
- B delegates their power to C
- C delegates to themselves

All 3 have a balance of 1M. Is C's voting power 3M or 2M?

If it's 3M, then if B clears their delegation, we will remove 1M from C's power but we should be removing 2M, right? 

If it's 2M, does B now still have a voting power of 1M? Do we currently allow them to vote in that case?

Reply:
> For my understanding, what happens in this case:
> 
> * wallet A delegates their voting power to wallet B
> * B delegates their power to C
> * C delegates to themselves

tracking so far.

> All 3 have a balance of 1M. Is C's voting power 3M or 2M?

C's voting power is 2M, vote power delegation is not transitive.

> If it's 3M, then if B clears their delegation, we will remove 1M from C's power but we should be removing 2M, right?
> 
> If it's 2M, does B now still have a voting power of 1M? Do we currently allow them to vote in that case?

B has a voting power of 1M, correct. My expectation is that they would be able to vote, since they have some voting power.

Tracing through this final state in your scenario:

* If B clears their delegation at this point, and...
* A has delegated their 1M balance to B
* C's balance and votes would both be 1M, their own self-delegated balance
* B's balance is 1M and votes are 1M, since they have not delegated to themselves
* A's balance is 1M and votes are 0.

Reply:
@arshsingh I ensconced this scenario in a test to ensure my explanation was correct.

Reply:
@mrmemes-eth what about this one:

- A delegates to B

B hasn't delegated yet, but their voting power is 1M, right?

- B delegates to themselves

Now B has a balance of 1M but voting power of 2M. If B clears their delegation, we will set their voting power to 0 since getVotes will return 2M. But shouldn't B's voting power be 1M in that case?

Reply:
@arshsingh good catch! How about the updated version? This also still resolves the case with previously incorrect delegation amounts for anyone that happened to delegate while the `delegateBySig` bug was active.

Reply:
Yep, updated version looks good to me 👍